homeassistant/bootstrap.py

```diff
@@ -55,44 +56,27 @@ def async_setup_component(hass: core.HomeAssistant, domain: str,
 
     This method is a coroutine.
     """
-    if domain in hass.config.components:
-        _LOGGER.debug('Component %s already set up.', domain)
-        return True
+    setup_tasks = hass.data.get(DATA_SETUP)
+
+    if setup_tasks is not None and domain in setup_tasks:
+        result = yield from setup_tasks[domain]
+        return result
 
     if not loader.PREPARED:
         yield from hass.loop.run_in_executor(None, loader.prepare, hass)
 
     if config is None:
         config = {}
 
-    components = loader.load_order_component(domain)
-
-    # OrderedSet is empty if component or dependencies could not be resolved
-    if not components:
-        _async_persistent_notification(hass, domain, True)
-        return False
-
-    setup_events = hass.data.get(DATA_SETUP)
-    if setup_events is None:
+    if setup_tasks is None:
         hass.data[DATA_PLATFORM] = {}
-        setup_events = hass.data[DATA_SETUP] = {}
+        setup_tasks = hass.data[DATA_SETUP] = {}
 
-    tasks = []
-    for component in components:
-        if component not in setup_events:
-            setup_events[component] = hass.async_add_job(
-                _async_setup_component(hass, component, config))
-        tasks.append(setup_events[component])
-
-    if tasks:
-        results = yield from asyncio.gather(*tasks, loop=hass.loop)
-        for idx, res in enumerate(results):
-            if not res:
-                _LOGGER.error('Component %s failed to setup', components[idx])
-                _async_persistent_notification(hass, components[idx], True)
-                return False
+    task = setup_tasks[domain] = hass.async_add_job(
```

Comment:
A task is a scheduled coroutine on the event loop that will only be executed once. No matter how many times you yield from it.

The first time we call `async_setup_component` with for example 'switch', we will see that there is no 'switch' task yet in `hass.data`. Thus we create one on this line.

The next time someone calls this to setup 'switch', for example some discovery, it will get the exact same Task object back. So if setup is done, it will get True and can continue. If setup is not done yet, it will wait till it is done.